syntax = "proto3";

package posts;

service PostService {
  rpc CreatePost (CreatePostRequest) returns (PostResponse);
  
  rpc GetPost (GetPostRequest) returns (PostResponse);
  
  rpc UpdatePost (UpdatePostRequest) returns (PostResponse);
  
  rpc DeletePost (DeletePostRequest) returns (DeletePostResponse);
  
  rpc ListPosts (ListPostsRequest) returns (ListPostsResponse);

  rpc ViewPost (ViewPostRequest) returns (ViewPostResponse);
  
  rpc LikePost (LikePostRequest) returns (LikePostResponse);
  
  rpc CreateComment (CreateCommentRequest) returns (CommentResponse);
  
  rpc ListComments (ListCommentsRequest) returns (ListCommentsResponse);
}

message CreatePostRequest {
  string title = 1;
  string description = 2;
  int32 user_id = 3;
  bool is_private = 4;
  repeated string tags = 5;
}

message GetPostRequest {
  int32 post_id = 1;
  int32 user_id = 2;
}

message UpdatePostRequest {
  int32 post_id = 1;
  int32 user_id = 2;
  string title = 3;
  string description = 4;
  bool is_private = 5;
  repeated string tags = 6;
}

message DeletePostRequest {
  int32 post_id = 1;
  int32 user_id = 2;
}

message DeletePostResponse {
  bool success = 1;
  string message = 2;
}

message ListPostsRequest {
  int32 page = 1;
  int32 per_page = 2;
  int32 user_id = 3;
  string tag = 4;
}

message ListPostsResponse {
  repeated Post posts = 1;
  int32 total_count = 2;
  int32 page = 3;
  int32 total_pages = 4;
}

message Post {
  int32 id = 1;
  string title = 2;
  string description = 3;
  int32 user_id = 4;
  bool is_private = 5;
  repeated string tags = 6;
  string created_at = 7;
  string updated_at = 8;
}

message PostResponse {
  Post post = 1;
  string error = 2;
}

message ViewPostRequest {
  int32 post_id = 1;
  int32 user_id = 2;
}

message ViewPostResponse {
  bool success = 1;
  string message = 2;
}

message LikePostRequest {
  int32 post_id = 1;
  int32 user_id = 2;
}

message LikePostResponse {
  bool success = 1;
  string message = 2;
}

message CreateCommentRequest {
  int32 post_id = 1;
  int32 user_id = 2;
  string content = 3;
}

message CommentResponse {
  Comment comment = 1;
  string error = 2;
}

message Comment {
  int32 id = 1;
  int32 post_id = 2;
  int32 user_id = 3;
  string content = 4;
  string created_at = 5;
}

message ListCommentsRequest {
  int32 post_id = 1;
  int32 page = 2;
  int32 per_page = 3;
}

message ListCommentsResponse {
  repeated Comment comments = 1;
  int32 total_count = 2;
  int32 page = 3;
  int32 total_pages = 4;
}
